sw/device/lib/msg_print.c

```diff
@@ -0,0 +1,107 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/msg_print.h"
+
+#include <stdarg.h>
+
+static void print_digit(print_char_func print_char, unsigned int d) {
+  print_char("0123456789ABCDEF"[d]);
+}
+
+static void print_num(print_char_func print_char, int w, unsigned int n,
+                      unsigned int base) {
+  if (--w > 0 || n >= base) {
+    print_num(print_char, w, n / base, base);
+  }
+  print_digit(print_char, n % base);
+}
+
+void msg_print(print_char_func print_char, const char *fmt, ...) {
+  char ch;
+  va_list va;
+
+  va_start(va, fmt);
+  while ((ch = *fmt) != '\0') {
```

Comment:
Assignments in loops... aren't the best. Why not
```c
for (; *fmt != '\0'; ++fmt) {
  char ch = *fmt;
  // ...
}
```
This also helps keep `ch` scoped to where it is used.

Reply:
Done